File: step_09_stack_queues/9.1_learning/P003_implement_stack_using_linked_list/Solution.java
/**
 * Problem: Implement Stack using Linked List
 * Difficulty: EASY | XP: 10
 *
 * Implement stack using singly linked list.
 * Head of the list = top of the stack.
 *
 * @author DSA_Nova
 */
public class Solution {

    // ============================================================
    // NODE DEFINITION
    // ============================================================
    static class Node {
        int val;
        Node next;

        Node(int val) {
            this.val = val;
            this.next = null;
        }
    }

    // ============================================================
    // STACK USING SINGLY LINKED LIST
    // All operations: O(1) time
    // Space: O(n) overall for n elements
    // ============================================================
    static class LinkedListStack {
        private Node top;   // head of the list = top of stack
        private int size;

        LinkedListStack() {
            this.top = null;
            this.size = 0;
        }

        /**
         * Push element onto the stack.
         * Insert at head of linked list.
         * Time: O(1)
         */
        void push(int val) {
            Node newNode = new Node(val);
            newNode.next = top;
            top = newNode;
            size++;
        }

        /**
         * Remove and return the top element.
         * Delete head of linked list.
         * Time: O(1)
         */
        int pop() {
            if (top == null) {
                System.out.println("Stack Underflow! Cannot pop.");
                return -1;
            }
            int val = top.val;
            top = top.next;  // old top is now unreferenced (GC will collect it)
            size--;
            return val;
        }

        /**
         * Return top element without removing.
         * Time: O(1)
         */
        int peek() {
            if (top == null) {
                System.out.println("Stack is empty! Cannot peek.");
                return -1;
            }
            return top.val;
        }

        /**
         * Check if stack is empty.
         * Time: O(1)
         */
        boolean isEmpty() {
            return top == null;
        }

        /**
         * Return the number of elements.
         * Time: O(1)
         */
        int size() {
            return size;
        }

        /**
         * Print stack from top to bottom.
         */
        void printStack() {
            if (isEmpty()) {
                System.out.println("Stack: (empty)");
                return;
            }
            StringBuilder sb = new StringBuilder("Stack (top->bottom): [");
            Node current = top;
            while (current != null) {
                sb.append(current.val);
                if (current.next != null) sb.append(", ");
                current = current.next;
            }
            sb.append("]");
            System.out.println(sb.toString());
        }
    }

    public static void main(String[] args) {
        System.out.println("=== Implement Stack using Linked List ===\n");

        LinkedListStack stack = new LinkedListStack();

        // Push operations
        stack.push(10);
        stack.push(20);
        stack.push(30);
        stack.push(40);
        stack.printStack();  // [40, 30, 20, 10]

        // Peek
        System.out.println("Peek: " + stack.peek());       // 40
        System.out.println("Size: " + stack.size());        // 4

        // Pop operations
        System.out.println("Pop:  " + stack.pop());         // 40
        System.out.println("Pop:  " + stack.pop());         // 30
        stack.printStack();  // [20, 10]

        // Continue popping
        System.out.println("Pop:  " + stack.pop());         // 20
        System.out.println("Pop:  " + stack.pop());         // 10
        System.out.println("isEmpty: " + stack.isEmpty());  // true

        // Underflow test
        stack.pop();   // Underflow!
        stack.peek();  // Empty!

        // No overflow -- linked list grows dynamically
        System.out.println("\n--- No overflow (dynamic sizing) ---");
        LinkedListStack bigStack = new LinkedListStack();
        for (int i = 1; i <= 10; i++) {
            bigStack.push(i * 100);
        }
        System.out.println("Pushed 10 elements. Size: " + bigStack.size());
        bigStack.printStack();

        // Pop all
        System.out.println("\n--- Pop all ---");
        while (!bigStack.isEmpty()) {
            System.out.println("Pop: " + bigStack.pop());
        }
        System.out.println("isEmpty: " + bigStack.isEmpty());
    }
}

File: step_09_stack_queues/9.1_learning/P003_implement_stack_using_linked_list/solution.py
"""
Problem: Implement Stack using Linked List
Difficulty: EASY | XP: 10

Implement stack using singly linked list.
Head of the list = top of the stack.
"""
from typing import Optional


# ============================================================
# NODE DEFINITION
# ============================================================
class Node:
    def __init__(self, val: int = 0):
        self.val = val
        self.next: Optional[Node] = None


# ============================================================
# STACK USING SINGLY LINKED LIST
# All operations: O(1) time
# Space: O(n) overall for n elements
# ============================================================
class LinkedListStack:
    def __init__(self):
        self.top: Optional[Node] = None  # head = top of stack
        self._size: int = 0

    def push(self, val: int) -> None:
        """Push element onto stack. Insert at head. Time: O(1)."""
        new_node = Node(val)
        new_node.next = self.top
        self.top = new_node
        self._size += 1

    def pop(self) -> int:
        """Remove and return top element. Delete head. Time: O(1)."""
        if self.top is None:
            print("Stack Underflow! Cannot pop.")
            return -1
        val = self.top.val
        self.top = self.top.next
        self._size -= 1
        return val

    def peek(self) -> int:
        """Return top element without removing. Time: O(1)."""
        if self.top is None:
            print("Stack is empty! Cannot peek.")
            return -1
        return self.top.val

    def is_empty(self) -> bool:
        """Check if stack is empty. Time: O(1)."""
        return self.top is None

    def size(self) -> int:
        """Return number of elements. Time: O(1)."""
        return self._size

    def print_stack(self) -> None:
        """Print stack from top to bottom."""
        if self.is_empty():
            print("Stack: (empty)")
            return
        elements = []
        current = self.top
        while current:
            elements.append(str(current.val))
            current = current.next
        print(f"Stack (top->bottom): [{', '.join(elements)}]")


if __name__ == "__main__":
    print("=== Implement Stack using Linked List ===\n")

    stack = LinkedListStack()

    # Push operations
    stack.push(10)
    stack.push(20)
    stack.push(30)
    stack.push(40)
    stack.print_stack()  # [40, 30, 20, 10]

    # Peek
    print(f"Peek: {stack.peek()}")        # 40
    print(f"Size: {stack.size()}")        # 4

    # Pop operations
    print(f"Pop:  {stack.pop()}")         # 40
    print(f"Pop:  {stack.pop()}")         # 30
    stack.print_stack()  # [20, 10]

    # Continue popping
    print(f"Pop:  {stack.pop()}")         # 20
    print(f"Pop:  {stack.pop()}")         # 10
    print(f"isEmpty: {stack.is_empty()}")  # True

    # Underflow test
    stack.pop()   # Underflow!
    stack.peek()  # Empty!

    # No overflow -- linked list grows dynamically
    print("\n--- No overflow (dynamic sizing) ---")
    big_stack = LinkedListStack()
    for i in range(1, 11):
        big_stack.push(i * 100)
    print(f"Pushed 10 elements. Size: {big_stack.size()}")
    big_stack.print_stack()

    # Pop all
    print("\n--- Pop all ---")
    while not big_stack.is_empty():
        print(f"Pop: {big_stack.pop()}")
    print(f"isEmpty: {big_stack.is_empty()}")
